x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/job/persistence/JobStorageDeletionTask.java

```diff
@@ -106,6 +113,72 @@ public void delete(String jobId, Client client, ClusterState state,
                 },
                 failureHandler);
 
+        // Step 4(c). If we have any hits, that means we are NOT the only job on this index, and should not delete it
+        ActionListener<SearchResponse> customIndexSearchHandler = ActionListener.wrap(
+            searchResponse -> {
+                if (searchResponse.getHits().totalHits > 0) {
+                    deleteByQueryExecutor.onResponse(true);
+                } else {
+                    logger.info("Running DELETE Index on [" + indexName + "] for job [" + jobId + "]");
+                    DeleteIndexRequest request = new DeleteIndexRequest(indexName);
+                    request.indicesOptions(IndicesOptions.lenientExpandOpen());
+                    executeAsyncWithOrigin(
+                        client.threadPool().getThreadContext(),
+                        ML_ORIGIN,
+                        request,
+                        ActionListener.<AcknowledgedResponse>wrap(
+                            deleteIndexResponse -> deleteAliases(jobId, client, deleteAliasHandler), //skip DBQ
+                            failureHandler),
+                        client.admin().indices()::delete);
+                }
+            },
+            failureHandler
+        );
+
+        // Step 4(b). If we are shared, skip index deletion. If not, determine if we are the only job on the custom index
+        ActionListener<GetAliasesResponse> getAliasesHandler = ActionListener.wrap(
+            aliasesResponse -> {
+                boolean onlySingleJob = true;
+                for (AliasMetaData aliasMetaData : aliasesResponse.getAliases().getOrDefault(indexName, new ArrayList<>())) {
+                    onlySingleJob = onlySingleJob && aliasMetaData.alias().endsWith(jobId); // circuit break for performance
+                }
+                if (onlySingleJob) { // Could we just skip this search and simply rely on the index suffixes?
```

Comment:
Do we trust our aliases enough to not do the additional search on the index?

The reason for the alias search in the first place is for when there are jobs on the same shared custom index where one does not have any documents yet.

Reply:
I think we should consider only doing the search. It'd be a search where we look for docs where the `job_id` isn't the one we're deleting and we set the `size` to `1`, which means it's quite a light-weight search. If there's another job created using the same custom index and has no docs, even if we delete the index, it is going to be recreated from the template as soon as results are written out. Could you confirm that @droberts195 ?

Reply:
> If there's another job created using the same custom index and has no docs, even if we delete the index, it is going to be recreated from the template as soon as results are written out.

Yes, that's correct.

I agree it's nice to only do one of the checks.  Let's start simple and just do the search.